core/src/main/java/io/parsingdata/metal/expression/value/Scope.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2013-2019 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value;
+
+import static java.math.BigInteger.ZERO;
+
+import static io.parsingdata.metal.Trampoline.complete;
+import static io.parsingdata.metal.Trampoline.intermediate;
+import static io.parsingdata.metal.Util.checkNotNull;
+import static io.parsingdata.metal.expression.value.NotAValue.NOT_A_VALUE;
+
+import java.util.Objects;
+
+import io.parsingdata.metal.Trampoline;
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseGraph;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.token.Token;
+
+/**
+ * A {@link ValueExpression} that prunes the {@link ParseState#order} field to
+ * let a nested {@link ValueExpression} have only a scoped view.
+ * <p>
+ * A Scope expression has two operands: <code>scopedValueExpression</code> (a
+ * {@link ValueExpression} and <code>scopeSize</code> (a
+ * {@link SingleValueExpression}). The <code>scopeSize</code> operand is first
+ * evaluated, if it evaluates to <code>NOT_A_VALUE</code> or a negative value
+ * an {@link IllegalArgumentException} is thrown. Otherwise the
```

Comment:
I feel the doc does talk quite a bit in terms of how it is implemented (which could of course also change).  It is probably good here where the token implementation resides, but I would expect at the point where people use it (i.e. the Shorthand), it would describe more about what it does, what the use case is and how you use it, with perhaps an example of how it can be used. But since there is no doc in the Shorthand yet, perhaps add here (though ideally Shorthand :smile:)? Because I think it would make it more clear how to use this expression, especially for people not familiar with  implementation details of Metal.

Reply:
I totally agree that we need documentation that describes _what_ things do and _how_ to use them in practice. I'm not entirely sure we should make `Shorthand` even more unwieldy by adding a massive amount of Javadoc to it.

I think that what we could use best is just some user guide that explains what all the Tokens/Expressions/ValueExpressions do in combination with 1-2 practical examples for everything. Let's first discuss what we need :)